fragment.go

```diff
@@ -1508,36 +1491,83 @@ func (f *fragment) bulkImportStandard(rowIDs, columnIDs []uint64, options *Impor
+	defer f.mu.Unlock()
 	if options.Clear {
-		changedN, err = localBitmap.RemoveN(positions...) // TODO benchmark AddN behavior with sorted/unsorted positions
+		err = f.importPositions(nil, positions, rowSet)
 	} else {
-		changedN, err = localBitmap.AddN(positions...) // TODO benchmark RemoveN behavior with sorted/unsorted positions
+		err = f.importPositions(positions, nil, rowSet)
 	}
+	return errors.Wrap(err, "bulkImportStandard")
+}
+
+// importPositions takes slices of positions within the fragment to set and
+// clear in storage. One must also pass in the set of unique rows which are
+// affected by the set and clear operations. It is unprotected (f.mu must be
+// locked when calling it). No position should appear in both set and clear.
+//
+// importPositions tries to intelligently decide whether or not to do a full
+// snapshot of the fragment or just do in-memory updates while appending
+// operations to the op log.
+func (f *fragment) importPositions(set, clear []uint64, rowSet map[uint64]struct{}) error {
+	var setBitmap *roaring.Bitmap
+	var clearBitmap *roaring.Bitmap
+	smallWrite := false
+	if len(set)+len(clear)+f.opN < f.MaxOpN {
+		smallWrite = true
+		// If the number of operations is small, we'll write directly to local storage
+		setBitmap = f.storage
+		clearBitmap = f.storage
+	} else {
+		// Create a temporary bitmap which will be populated by rowIDs and columnIDs
+		// and then merged into the existing fragment's bitmap.
+		setBitmap = roaring.NewBitmap()
+		clearBitmap = roaring.NewBitmap()
+
+		// Disconnect op writer so we don't append updates.
+		setBitmap.OpWriter = nil
+		clearBitmap.OpWriter = nil
+	}
+
+	changedN, err := setBitmap.AddN(set...) // TODO benchmark Add/RemoveN behavior with sorted/unsorted positions
 	if err != nil {
 		return errors.Wrap(err, "adding positions")
 	}
-	f.stats.Count("ImportBit", int64(changedN), 1)
-	f.opN += changedN
 
-	var results *roaring.Bitmap
-	if !smallWrite {
-		// Merge localBitmap into fragment's existing data.
-		if options.Clear {
-			if f.storage.Count() > 0 {
-				results = f.storage.Difference(localBitmap)
-			} else {
-				results = roaring.NewBitmap()
-			}
-		} else {
-			if f.storage.Count() > 0 {
-				results = f.storage.Union(localBitmap)
-			} else {
-				results = localBitmap
-			}
-		}
+	if smallWrite {
+		f.stats.Count("ImportBit", int64(changedN), 1)
+		f.opN += changedN
+		changedN, err = clearBitmap.RemoveN(clear...)
+		f.stats.Count("ClearBit", int64(changedN), 1)
+		f.opN += changedN
 	} else {
+		// when doing big imports, we're going to difference the local bitmap,
+		// so we need to add rather than removing. TODO: figure out why we
+		// aren't adding/removing directly from storage.
+		_, err = clearBitmap.AddN(clear...)
+	}
+	if err != nil {
+		return errors.Wrap(err, "clearing positions")
+	}
+
+	var results *roaring.Bitmap
+	if smallWrite {
 		results = f.storage
+	} else if beforeCnt := f.storage.Count(); !smallWrite && beforeCnt > 0 {
```

Comment:
I don't think we need the explicit `!smallWrite` here because this is the else from a test on smallWrite.

See also the previous question about an Any() operation.